Migrate codebase to typescript

ankita240796 opened the pull request:
## Explanation

This PR introduces typescript in codebase and fixes #6343. 

Changes:

- Rename js to ts. The protractor test are still in js and can be moved to ts once Angular2 comes in. This is because protractor types are installed by default and we need to use import statements to import from the types (as specified in the examples [here](https://github.com/angular/protractor/blob/5.4.1/exampleTypescript/angularPage.ts)). The compiled javascript is not compatible with AngularJs

- Karma test are migrated to typescript and all test are passing. Backend test and e2e test also pass.

- Update the build process
   - Build process before typescript:

   ![55835884_342453749715982_7503872375138549760_n](https://user-images.githubusercontent.com/15226041/55265122-e2812100-529c-11e9-9ada-a86cf2ffcb59.jpg)

   - Build process after typescript:

   ![55752529_1927364670725445_8401529550843936768_n](https://user-images.githubusercontent.com/15226041/55265932-6d174f80-52a0-11e9-8b4b-05b448a6cac7.jpg)

   - A new compiled_js_dir is added from where js files are minified and copied. The hashes are computed for these js files. So, we need to compile each time before starting the build tasks. 

   - The issue here is to reduce compilation time. The fix can be to have a readme in the local_compiled_js dir with the last commit for which compilation was done. Then we can find the diff between the last and the current commit and compile only the changed files. This requires adding some custom type definitions ([Ref](https://github.com/oppia/oppia/commit/792c533f9127b4aa287f8704068260b6e2e092bb#diff-6badab1552630db8553344d0f39ef2dd)) to avoid compilation errors.

- Adding typescript-eslint parser. This has resulted in following issues:
   - Lint is working locally when I run lint for some of the files or some of the folders but fails working when I run it for all the files. Once the js and ts linting ends, there is no output and the circleci build times out (Same happens locally)
   - Excluding the generated files (`extensions/interactions/LogicProof/static/js/generatedDefaultData.ts`, `extensions/interactions/LogicProof/static/js/generatedParser.ts`, `core/templates/dev/head/expressions/ExpressionParserService.ts`) fixes the no output issue but the linter exits with code 1 after the js and ts linting is done.
 I am unable to find a reason behind this. @apb7 may you please guide me here, Thanks!

ankita240796 committed:
Fix type issues

ankita240796 committed:
Rename files

ankita240796 committed:
Disable lint

ankita240796 committed:
Enable lint

ankita240796 committed:
Rename test files, disable lint

ankita240796 committed:
Enable lint

ankita240796 committed:
Add ajv-keywords to avoid conflict

ankita240796 committed:
Fix some lint issues, disable lint

ankita240796 committed:
Enable lint

ankita240796 committed:
Fix version issue

ankita240796 committed:
Revert version

seanlip commented:
/cc @oppia/dev-workflow-team Just wanted to flag the linter problem that @ankita240796 mentioned in her first comment on this thread.

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into migration-ts

ankita240796 committed:
Disable lint

seanlip commented on core/templates/dev/head/pages/exploration_editor/translation_tab/AudioTranslationBarDirective.js:
```diff
@@ -48,9 +48,9 @@ oppia.directive('audioTranslationBar', [
-            scope.dropAreaIsAccessible) {
-            files = evt.originalEvent.dataTransfer.files;
+          if ((<Element><any>evt.target).classList.contains(
+            'oppia-drop-area-message')  && scope.dropAreaIsAccessible) {
```
Drop extra space before `&&`.

/cc @oppia/dev-workflow-team -- missing lint check?

ankita240796 replied:
Lint is actually not working on this PR due to which this is missed. 

Done.

ankita240796 committed:
Enable lint

seanlip commented on extensions/interactions/LogicProof/static/js/sharedSpec.js:
```diff
@@ -17,7 +17,8 @@
  */
 
-var errorWrapper = function(dubiousFunction, input, parameter = null) {
+var errorWrapperForSharedSpec = function(
```
I think we need to find a way to not have this be in global scope.

ankita240796 replied:
https://stackoverflow.com/a/19263750 - Can we use the second way to avoid polluting the global scope?

seanlip replied:
Hm, that's a bit indirect.

Here's another idea: I believe you can have nested describe blocks. How about having two layers of describe blocks, and declare this in the outer one?

ankita240796 replied:
Sounds good, will do.

ankita240796 replied:
Done.

seanlip commented on feconf.py:
```diff
@@ -61,8 +61,14 @@
 VALUE_GENERATORS_DIR = os.path.join('extensions', 'value_generators')
-VISUALIZATIONS_DIR = os.path.join('extensions', 'visualizations')
+
+VISUALIZATIONS_DIR = os.path.join(
```
For consistency, should this be renamed VISUALIZATIONS_DIR_FOR_JS?

(Also, are non-JS files in extensions/visualizations not used?)

ankita240796 replied:
No other files are not used in `visualization_registry.py`. The HTML templates are used in the directives itself. 

Done.

seanlip commented on scripts/build.py:
```diff
@@ -263,14 +266,19 @@ def _compare_file_count(first_dir_path, second_dir_path):
 
     Args:
-       first_dir_path: str. First directory to compare.
+       first_dir_path: str/list. First directory to compare.
```
We use | not /, and also there is missing typeinfo for the list elements.

That said, this seems conceptually wrong. A dirpath is a string. It should not also be a list.

ankita240796 replied:
Added new function.

seanlip commented on typings/custom-oppia-defs.d.ts:
```diff
@@ -0,0 +1,22 @@
+// These declarations are required to avoid compile errors while compiling
```
We should talk about this, I don't know what the criterion is for files to go in here, and this means it cannot be enforced.

ankita240796 replied:
This will be removed by using typescript in watch mode.

seanlip commented on typings/third-party-defs.d.ts:
```diff
@@ -3,5 +3,6 @@ declare var MIDI: any;
 declare var Guppy: any;
 declare var Sk: any;
 declare var safari: any;
```
Drop safari and system (see earlier comment in #6441)

ankita240796 replied:
Done.

ankita240796 committed:
Drop extra space

ankita240796 committed:
Disable lint

ankita240796 committed:
Add visualization dir for js for consistency

ankita240796 committed:
Rename generated files to js

ankita240796 committed:
Add new function for comparing count against list of dirs

ankita240796 committed:
Remove stuff from global scope

ankita240796 committed:
Add comments for typings

ankita240796 committed:
Remove unused functions - fix system and safari

ankita240796 committed:
Remove unused function, use spread operator

ankita240796 committed:
Enable lint

ankita240796 commented:
Hi @seanlip, I have addressed the review comments and made the changes as discussed in the meeting. I am still working on typescript watch mode and will update that once I am done. PTAL at the other changes, Thanks!

ankita240796 committed:
Temporary change to fix e2e test

ankita240796 committed:
Temporary order change to fix e2e test

ankita240796 commented:
/cc @oppia/dev-workflow-team The lint test should not pass on this PR but they are passing strangely. If I run test for some small number of files locally, errors are caught but for the complete codebase, the linter stops once js and ts linting is done. Can you please look into this, Thanks!

ankita240796 committed:
Revert changes

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into migration-ts

ankita240796 committed:
Disable lint

ankita240796 committed:
Enable lint

seanlip commented on scripts/build.py:
```diff
@@ -287,6 +294,30 @@ def _compare_file_count(first_dir_path, second_dir_path):
 
+def _compare_file_count_for_multiple_dirs(
+        first_dir_list, second_dir_path):
+    """Ensure that two dir's file counts match.
```
This is a bit misleading, it should be "Ensure that the total count of files in all directories in the first list matches the count of files in the second directory."

Alternatively, you could generalize the existing function: compare_file_counts(first_dir_list, second_dir_list) and use 1-element lists as needed.

ankita240796 replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -529,7 +560,7 @@ def should_file_be_built(filepath):
     """
     if filepath.endswith('.js'):
-        return not any(
+        return not filepath in JS_FILEPATHS_NOT_TO_BUILD and not any(
```
"filepath not in" is more pythonic, I think

ankita240796 replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -1026,17 +1059,23 @@ def _verify_build(input_dirnames, output_dirnames, file_hashes):
         Args:
-            input_dirnames: list(str). List of directory paths that contain
-                source files.
+            input_dirnames: list(str|list(str)). List of directory paths that
```
No, don't introduce type polymorphism here, for the same reasons as before.

ankita240796 replied:
Updated.

seanlip commented on scripts/build.py:
```diff
@@ -1026,17 +1059,23 @@ def _verify_build(input_dirnames, output_dirnames, file_hashes):
+                contain source files.
             output_dirnames: list(str). List of directory paths that contain
                 built files.
+            input_dirnames_type: Type of input_dirnames - whether it is a single
```
No, don't do this. Find a way to break things down that makes things clearer (see example above), don't do type magic via additional args. Having to do if/else for stuff like this is an antipattern and makes the code harder to maintain.

ankita240796 replied:
Updated.

seanlip commented on scripts/build.py:
```diff
@@ -1137,7 +1176,11 @@ def generate_build_directory():
-    _verify_build(SOURCE_DIRS, COPY_OUTPUT_DIRS, hashes)
+    # This is required since we need to compare build/extensions and
+    # build/templates against both the compiled js dir and dev dir.
+    SOURCE_DIRS_TYPE = [
```
Nope -- as above, find another way that reads more cleanly.

ankita240796 replied:
Updated.

ankita240796 committed:
Update comparison of file counts

ankita240796 committed:
Update comments

ankita240796 commented:
Hi @seanlip, I have updated the method to verify file count, PTAL. Thanks!

ankita240796 committed:
Enable lint

ankita240796 committed:
Debug extra file in build

ankita240796 committed:
Move print before error is raised

seanlip commented on scripts/build.py:
```diff
@@ -295,29 +295,52 @@ def _compare_file_count(first_dir_path, second_dir_path):
+            count as the target directory list.
     """
 
     first_dir_file_count = 0
```
first_dir_list_file_count or first_dirs_file_count?

seanlip replied:
Better still, use file_counts = [0, 0] and put a for loop around the dirwalk code, so that you don't repeat code.

ankita240796 replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -295,29 +295,52 @@ def _compare_file_count(first_dir_path, second_dir_path):
     """
 
     first_dir_file_count = 0
+    files1 = []
```
filepaths_in_first_dirs

ankita240796 replied:
This was just to debug a commit. It is removed now.

seanlip commented on scripts/build.py:
```diff
@@ -295,29 +295,52 @@ def _compare_file_count(first_dir_path, second_dir_path):
+                        files1.append(filepath)
+
+    second_dir_file_count = 0
+    files2 = []
```
filepaths_in_second_dirs

ankita240796 replied:
This was just to debug a commit. It is removed now.

seanlip commented on scripts/build.py:
```diff
@@ -295,29 +295,52 @@ def _compare_file_count(first_dir_path, second_dir_path):
+
+    second_dir_file_count = 0
+    files2 = []
+    for second_dir_path in second_dir_list:
```
Remove the repetition. See comment above.

ankita240796 replied:
I have added file_counts but I am not sure how to avoid having two loops for each dir since both have different number of elements for extensions and templates.

seanlip commented on scripts/build.py:
```diff
@@ -1109,6 +1118,42 @@ def _verify_build(
+def _verify_file_count(
+        input_dirnames, output_dirnames):
+    """Verify number of files between source directory and final directory
+        matches after build process finishes.
```
Deindent by 4.

ankita240796 replied:
This function is removed now.

seanlip commented on scripts/build.py:
```diff
@@ -1109,6 +1118,42 @@ def _verify_build(
+    """Verify number of files between source directory and final directory
+        matches after build process finishes.
+
+        Args:
```
Deindent all this by 4.

ankita240796 replied:
This function is removed now.

seanlip commented on scripts/build.py:
```diff
@@ -1109,6 +1118,42 @@ def _verify_build(
+                source files.
+            output_dirnames: list(str). List of directory paths that contain
+                built files.
+    """
```
Missing Raises: docstring

ankita240796 replied:
No exception is raised by this function, it is raised by compare file count function which has the docstring.

seanlip replied:
True, but the whole point of this function (semantically) is to raise an exception if there's a problem, so you might want to document it in this case.

ankita240796 replied:
This function is now removed.

seanlip commented on scripts/build.py:
```diff
@@ -1109,6 +1118,42 @@ def _verify_build(
+    list and all directories in final directory list matches after build
+    process finishes.
+
+        Args:
```
indentation

ankita240796 replied:
Function is removed now.

seanlip commented on scripts/build.py:
```diff
@@ -1109,6 +1118,42 @@ def _verify_build(
+    process finishes.
+
+        Args:
+            input_dirnames_list: list(list(str)). List of list of directory
```
why not just list(str)? Isn't it a list of dirnames?

Ditto below.

ankita240796 replied:
Updated.

seanlip commented on scripts/build.py:
```diff
@@ -1167,20 +1212,24 @@ def generate_build_directory():
+    OUTPUT_DIRS = [ASSETS_OUT_DIR, THIRD_PARTY_GENERATED_OUT_DIR]
+    _verify_file_count(SOURCE_DIRS, OUTPUT_DIRS)
+
+    SOURCE_DIRS_LIST = [[
```
This is getting confusing TBH.

Why not call _verify_file_count twice and have it operate on two dir lists? Once you start doing lists of lists of dirs it becomes confusing as the lists in the middle layer don't have a good conceptual meaning.

ankita240796 replied:
Updated - I split the source and output into four lists and called compare_file_count for each pair.

seanlip commented on scripts/build_test.py:
```diff
@@ -176,13 +176,14 @@ def test_compare_file_count(self):
         build.safe_delete_directory_tree(EMPTY_DIR)
 
-    def test_compare_file_count_with_multiple_dirs(self):
+    def test_compare_file_count_for_multiple_dirs(self):
```
You have made a lot of changes and additions to the code in build.py but the changes to the tests here are fairly minimal. Are you sure these tests are covering all the possible cases?

ankita240796 replied:
Now there is only one function and it is covered in testing.

ankita240796 committed:
Revert debug changes

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into migration-ts

ankita240796 committed:
Rename files

ankita240796 committed:
Address review comments

ankita240796 committed:
Disable lint

ankita240796 committed:
Enable lint

ankita240796 commented:
Hi @seanlip, I have addressed the review comments, PTAL. Thanks!

ankita240796 committed:
Remove duplicate code

seanlip reviewed (changes requested):
Thanks @ankita240796, I see no further issues!

However, I tried to check out this branch on my system and found that start.sh no longer works for me. I get:

```
 npm ERR! peer dep missing: ajv@^5.0.0, required by ajv-keywords@2.1.1
```

We need to debug this and similar issues before we can merge this, since it is going to impact developers' workflow. Could you please write here the full text of the instructions that you are planning to send to developers? I think that I, as well as someone from @oppia/dev-workflow-team, should checkout your branch and do a test run before we merge this -- it should work out of the box.

Also /cc @vojtechjelinek -- if you have time, @vojtechjelinek, could you PTAL at just the build scripts and sign off on them? Thanks!

ankita240796 commented:
Hi @seanlip, the full text of the instructions for the developers is as follows:

- Remove `ajv`, `ajv-keywords` and `eslint` folder from `Parent directory of oppia/node_modules`
- Run `bash scripts/install_third_party.sh`

No other steps are required for the current PR for development setup(Some steps will be added once I add watch mode compilation). Apart from that, I will need to share a wiki link which adds the basic coding guide for typescript. Though the compiler will raise errors if code is invalid but I will prepare a basic wiki guide before this PR gets in to ensure that developers are familiar with the basics.

Apart from that I would suggest @YashJipkate @lilithxxx @anubhavsinha98 @apb7 to see why the lint test are passing here though I get some failures when I run test on individual folders.

I have tested the PR locally and in prod mode but it would be better if we test the functionalities as done by the release testing team once again before the PR gets in. @nithusha21 What do you suggest?

I am a bit busy with my proposal but I will add the typescript watch mode stuff and wiki guide latest by this Thursday.

Thanks!

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into migration-ts

ankita240796 committed:
Rename new files to ts

ankita240796 committed:
Remove extra typings

ankita240796 committed:
Disable lint

ankita240796 committed:
Enable lint

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into migration-ts

ankita240796 committed:
Fix codeowners issue

ankita240796 committed:
Disable lint

ankita240796 committed:
Enable lint

ankita240796 committed:
Add issue link to jquery ui typings

ankita240796 commented:
Hi @seanlip, here is the wiki update (to the coding guidelines page) which I am planning to make once this PR goes in:

## Typescript
- Make sure to follow all the javascript rules here as well.
- Declare a variable before usage. All loop variables should also be declared.
- Do not add new properties to a declared variable. Ensure that all properties are declared in the variable declaration.
- Always initialize a variable at declaration. If you do not want a specific value at declaration, initialize the variable with a null value.
- Do not overwrite the variable with a different type. Instead create a new variable whenever you have a different use case.
- If you get compilation error which says that a property does not exist on a particular type, go through the type definitions of the type and do a type casting if required. You can add a new custom type definitions if type casting is not possible.

### Karma test specific guidelines
- Use `angular.mock.module` instead of `module` since the typings for angular-mocks does not support the usage of module.
- Use `angular.mock.inject` instead of `inject` to maintain a consistent behaviour.

### When to add custom type defintions to the typings folder?
- If you find a missing property in a typings package, create an issue [here](https://github.com/DefinitelyTyped/DefinitelyTyped) and a new file for the custom types with the issue link in the top of the file.
- If you add a package for which type definitions are not found [here](https://github.com/DefinitelyTyped/DefinitelyTyped), add it to `third-party-defs.d.ts`
- If you add a new property on window which is not present in typings for window, add it to `custom-window-defs.d.ts`
- If you add a property on scope defined in a link function, add it to `custom-scope-defs.d.ts` and add a comment specifying the filename for which it is added.
- Make sure that all files have relevant comments. Go through the existing files and try to follow the same pattern when adding a new file.

PTAL, Thanks!

seanlip commented:
Thanks, @ankita240796! Some notes:

- Suggest including code snippets with your guidelines -- i.e.  examples of what to do vs what not to do.
- Suggest making yourself codeowner for the *.d.ts files.
- "Make sure that all files have relevant comments." -- Not sure I understand this, maybe try to state more clearly what exactly the developer should do?

Otherwise, I think it looks good. Thanks!

ankita240796 committed:
Resolve conflicts

ankita240796 committed:
Fix lint issues, disable lint

ankita240796 committed:
Enable lint

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into migration-ts

ankita240796 committed:
Rename left out file

ankita240796 committed:
Merge branch 'migration-ts' of github.com:ankita240796/oppia into migration-ts

ankita240796 committed:
Disable lint

ankita240796 committed:
Become codeowner for typings

ankita240796 committed:
Remove extra any

ankita240796 committed:
Enable lint

ankita240796 commented:
Hi @seanlip, here is the updated wiki text according to your suggestions. I have also added myself as codeowner for typings files. PTAL, Thanks!

## Typescript
- Make sure to follow all the javascript rules here as well.
- Declare a variable before usage. For instance:
  
  **Wrong usage:**
  ```              
  exampleVar = true;
  if (someCondition) {
    exampleVar = false;
  }
  ```
  **Right usage:**
  ```              
  var exampleVar = true;
  if (someCondition) {
    exampleVar = false;
  }
  ```        
- All loop variables should be declared. For instance:     

  **Wrong usage:**
  ```              
  for (item in itemList) {
    ...
  }
  ```
  **Right usage:**
  ```              
  for (var item in itemList) {
    ...
  }
  ```    

- Do not add new properties to a declared variable. Ensure that all properties are declared in the variable declaration. For instance:

  **Wrong usage:**
  ```              
  var person = {
    name: 'name',
    age: 'age'
  };
  if (someCondition) {
    person.address = 'address';
  }
  ```
  **Right usage:**
  ```              
  var person = {
    name: 'name',
    age: 'age',
    address: null
  };
  if (someCondition) {
    person.address = 'address';
  }
  ``` 
- Always initialize a variable at declaration. If you do not want a specific value at declaration, initialize the variable with a null value. For instance:

  **Wrong usage:**
  ```              
  var person;
  if (someCondition) {
    person = 'name';
  }
  ```
  **Right usage:**
  ```              
  var person = null;
  if (someCondition) {
    person = 'name';
  }
  ``` 
- Do not overwrite the variable with a different type. Instead create a new variable whenever you have a different use case. For instance:

  **Wrong usage:**
  ```              
  var person = {
    name: 'name',
    schoolName: 'school name'
  };
  ...
  person = {
    name: 'name',
    officeName: 'office name'
  };
  ```
  **Right usage:**
  ```              
  var personForSchool = {
    name: 'name',
    schoolName: 'school name'
  };
  ...
  var personForOffice = {
    name: 'name',
    officeName: 'office name'
  };
  ``` 
- If you get compilation error which says that a property does not exist on a particular type, go through the type definitions of the type and do a type casting if required. For instance:
  ```              
  var checkMismatch = function(searchQuery) {
    var isMismatch = true;
    $('.oppia-search-bar-input').each(function(index) {
      if ((<string>$(this).val()).trim() === searchQuery) {
        isMismatch = false;
      }
    });
    return isMismatch;
  };
  ```
  Here `$(this).val()` is type casted to a string by using `<string>$(this).val()`
  If we do not use a typecast, typescript will give a error `Property 'trim' does not exist on type 'string | number | string[]'` since val can be a string or a number or a string array. So, to use trim we specifically need it as a string. 

  In many cases, you may also need to typecast to `<any>` first and then to the desired type. For example, for converting a number to string, you will need `<string><any>` because neither type sufficiently overlaps with the other.

  You can add a new custom type definitions if type casting is not possible. In the file `typings/custom-element-defs.d.ts`, we add a new property to `HTMLElement` by adding a custom type defintion. In this type casting cannot be used, since we are adding a new property to the existing type instead of changing it to some other type.

### Karma test specific guidelines
- Use `angular.mock.module` instead of `module` since the typings for angular-mocks does not support the usage of module.
- Use `angular.mock.inject` instead of `inject` to maintain a consistent behaviour.

### When to add custom type defintions to the typings folder?
- If you find a missing property in a typings package, create an issue [here](https://github.com/DefinitelyTyped/DefinitelyTyped) and a new file for the custom types with the issue link in the top of the file.
- If you add a package for which type definitions are not found [here](https://github.com/DefinitelyTyped/DefinitelyTyped), add it to `third-party-defs.d.ts`
- If you add a new property on window which is not present in typings for window, add it to `custom-window-defs.d.ts`
- If you add a property on scope defined in a link function, add it to `custom-scope-defs.d.ts` and add a comment specifying the filename for which it is added.
- Make sure that all files have comments which explain why these custom type defintions are required and additional comments to explain each new added property if required. For example, `typings/custom-scope-defs.d.ts` has a top level comment explaining that the type defintions are needed for properties defined on scope in link function and then there are additional comments with properties added specifying which file they belong to. Go through the existing files and try to follow the same pattern when adding a new file.

ankita240796 committed:
Add watch mode compilation

ankita240796 committed:
Merge branch 'migration-ts' of github.com:ankita240796/oppia into migration-ts

ankita240796 committed:
Disable lint

ankita240796 committed:
Enable lint

ankita240796 commented:
Hi @seanlip, I have added watch mode compilation for whenever user runs a development server locally but the changes are sometimes detected and sometimes not with a message that `There are too many folders, you may need to restart the development server to see the changes`. Can you please check if this happens on your system too. Thanks!

seanlip commented:
@ankita240796, that's an odd message. Are you sure it comes from typescript and not App Engine? I ask because I've seen something like that before with App Engine and I'm a bit surprised that typescript has the notion of a "development server".

For testing can you check with other folks on the Angular upgrade team first?

seanlip commented on scripts/build.py:
```diff
@@ -1177,6 +1177,40 @@ def compile_typescript_files(project_dir):
+                    # We are checking only the last line here since
+                    # whenever typescript is done with compilation
+                    # with or without errors, the last line will
+                    # always read Found x errors. Watching for file
```
Put quotation marks around the message.

ankita240796 replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -1177,6 +1177,40 @@ def compile_typescript_files(project_dir):
+                    last_output = lines[len(lines) - 1]
+                    if 'Watching for file changes' in last_output:
+                        return
+
```
Only two blank lines above top-level functions. /cc @oppia/dev-workflow-team

ankita240796 replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -1177,6 +1177,40 @@ def compile_typescript_files(project_dir):
+    print 'Compiling ts files in watch mode...'
+    cmd = '../node_modules/typescript/bin/tsc --project %s --watch' % (
+        project_dir)
+    subprocess.Popen('%s > tsc_output_log.txt' % cmd, shell=True)
```
I don't think you need shell=True here? Try to avoid using shell=True where possible, it has security implications.

ankita240796 replied:
I tried removing it but strangely on splitting the command, typescript gives an error that source files cannot be specified with the project dir option (tsc_output_log.txt is being treated as a source file)

seanlip replied:
That seems weird, I think it's worth looking into that further.

ankita240796 replied:
Sure, I am looking into it.

ankita240796 replied:
This is fixed now, PTAL.

seanlip commented on scripts/build.py:
```diff
@@ -1177,6 +1177,40 @@ def compile_typescript_files(project_dir):
     subprocess.check_call(cmd, shell=True)
 
 
+def compile_typescript_files_in_watch_mode(project_dir):
```
compile_typescript_files_continuously

and then, in the docstring, you can explain what "continuously" means.

ankita240796 replied:
Done.

seanlip reviewed (approved):
Hi @ankita240796, this looks pretty good, thanks! Only a few comments but basically LGTM. The proposed wiki instructions look good too.

Let's start the final process for getting this PR in. Could you do the following:

- Fix the remaining (small) comments.
- Provide the full updated text of the email you will be sending to oppia-dev@ as soon as this PR gets merged.
- Ask for a couple of Oppia developers to try out your branch (following the email in the previous step, with no further instructions) and see if they can get things running smoothly on their machine, and that changes they make to code reflect in their dev server. It would help if you provide specific test instructions (e.g. after you do this change and you reload this page, you should see XXX). They should then comment on this PR with the results of their test.
- Once all the above is done, let's merge this PR and then immediately put up the wiki page and send the email to oppia-dev@.

Thanks!

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into migration-ts

ankita240796 committed:
Address review comments

ankita240796 committed:
Merge branch 'migration-ts' of github.com:ankita240796/oppia into migration-ts

ankita240796 committed:
Disable lint

ankita240796 committed:
Enable lint

ankita240796 commented:
Hi @seanlip,

>  Are you sure it comes from typescript and not App Engine? I ask because I've seen something like that before with App Engine and I'm a bit surprised that typescript has the notion of a "development server".

Sorry for the confusion. I forgot to specify that the message comes from App Engine. Also, I have asked the Angular upgrade team to check the same on their system.

I have addressed all the review comments, PTAL.

Here is the updated text for the email to be sent to all the developers:

```
Hello all,

We are introducing typescript to the codebase for a smooth upgrade to Angular2 
from AngularJS. 
Here are some of the changes required to update your setup to use typescript:

- Update your develop branch
- Remove ajv, ajv-keywords and eslint from node_modules folder
- Run the command: bash scripts/install_third_party.sh
- Go through the coding guidelines for Typescript in the wiki here(wiki link)

Do let me know if you face any issues with this.

Thanks,
Ankita
```
If this mail looks fine, I will ask a couple of developers to test this PR with this mail and a set of instructions for what stuff to test.

PTAL, Thanks!

seanlip commented:
Ah, I think you can ignore the App Engine error. It's always been there and isn't a new thing.

The email looks good! I suggest you go ahead with the tests, and once you get confirmation, we can merge this (pending that last comment above).

ankita240796 committed:
Remove shell=True

ankita240796 committed:
Disable lint

ankita240796 committed:
Enable lint

ankita240796 commented:
Hi @seanlip, I have addressed the last comment. I have also sent an email to some of the developers asking if they will be able to help with the testing.

> Ah, I think you can ignore the App Engine error. It's always been there and isn't a new thing.

No, actually when I do a compilation without watch mode this message does not show up and AppEngine detects any change made manually to files in local_compiled_js.

On the other hand in watch mode compilation, this message shows up and AppEngine does not detect change when a recompiled js file is produced. I tried manually changing js files in local_compiled_js in this case but even that is not detected. Another thing which happens is that sometimes on starting the server, changes are detected in almost all files in local_compiled_js. I am not able to figure out why this is happening. May you please help me here, Thanks!

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into migration-ts

ankita240796 committed:
Merge branch 'migration-ts' of github.com:ankita240796/oppia into migration-ts

ankita240796 committed:
Disable lint

ankita240796 committed:
Fix errors

ankita240796 committed:
Enable lint

ankita240796 committed:
Resolve conflicts

ankita240796 committed:
Merge branch 'migration-ts' of github.com:ankita240796/oppia into migration-ts

ankita240796 committed:
Disable lint

ankita240796 committed:
Enable lint

ankita240796 committed:
Resolve conflict

ankita240796 committed:
Merge branch 'migration-ts' of github.com:ankita240796/oppia into migration-ts

ankita240796 committed:
Disable lint

ankita240796 committed:
Enable lint

ankita240796 committed:
Fix lint issue

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into migration-ts

ankita240796 committed:
Fix codeowners, disable lint

ankita240796 committed:
Enable lint

ankita240796 commented:
Hi @anubhavsinha98 @YashJipkate @apb7 @lilithxxx, the lint test is failing with the following message:

```
************* Module oppia.schema_utils
C:214,15: More than one statement on a single line (multiple-statements)
```

I have not made any change to schema_utils file as well as there is no statement on the pointed line [here](https://github.com/oppia/oppia/blob/develop/schema_utils.py#L214). I also ran the test locally on schema_utils.py and no error was detected. Can you all please look into this, Thanks!

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into migration-ts

ankita240796 committed:
Disable lint

ankita240796 committed:
Enable lint

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into migration-ts

ankita240796 committed:
Disable lint

ankita240796 committed:
Enable lint

brianrodri reviewed (approved):
LGTM! May I suggest a short set of instructions for contributors who wish to introduce new JS files (like me 🙃) so that we can help you avoid redoing this PR as we add new files?

kevinlee12 reviewed (approved):
lgtm!

ankita240796 commented:
> LGTM! May I suggest a short set of instructions for contributors who wish to introduce new JS files (like me ) so that we can help you avoid redoing this PR as we add new files?

Hi @brianrodri  That would be great! I think the rules [here](https://github.com/oppia/oppia/pull/6530#issuecomment-480583655) are sufficient since violation of these rules only can result in compilation errors.

bansalnitish reviewed (approved):
LGTM, Thanks @ankita240796!

brianrodri commented:
> Hi @seanlip, here is the updated wiki text according to your suggestions. I have also added myself as codeowner for typings files. PTAL, Thanks!
> 
> ## Typescript
> * Make sure to follow all the javascript rules here as well.
> * Declare a variable before usage. For instance:
>   **Wrong usage:**
>   ```
>   exampleVar = true;
>   if (someCondition) {
>     exampleVar = false;
>   }
>   ```
>   **Right usage:**
>   ```
>   var exampleVar = true;
>   if (someCondition) {
>     exampleVar = false;
>   }
>   ```
> * All loop variables should be declared. For instance:
>   **Wrong usage:**
>   ```
>   for (item in itemList) {
>     ...
>   }
>   ```
>   **Right usage:**
>   ```
>   for (var item in itemList) {
>     ...
>   }
>   ```
> * Do not add new properties to a declared variable. Ensure that all properties are declared in the variable declaration. For instance:
>   **Wrong usage:**
>   ```
>   var person = {
>     name: 'name',
>     age: 'age'
>   };
>   if (someCondition) {
>     person.address = 'address';
>   }
>   ```
>   **Right usage:**
>   ```
>   var person = {
>     name: 'name',
>     age: 'age',
>     address: null
>   };
>   if (someCondition) {
>     person.address = 'address';
>   }
>   ```
> * Always initialize a variable at declaration. If you do not want a specific value at declaration, initialize the variable with a null value. For instance:
>   **Wrong usage:**
>   ```
>   var person;
>   if (someCondition) {
>     person = 'name';
>   }
>   ```
>   **Right usage:**
>   ```
>   var person = null;
>   if (someCondition) {
>     person = 'name';
>   }
>   ```
> * Do not overwrite the variable with a different type. Instead create a new variable whenever you have a different use case. For instance:
>   **Wrong usage:**
>   ```
>   var person = {
>     name: 'name',
>     schoolName: 'school name'
>   };
>   ...
>   person = {
>     name: 'name',
>     officeName: 'office name'
>   };
>   ```
>   **Right usage:**
>   ```
>   var personForSchool = {
>     name: 'name',
>     schoolName: 'school name'
>   };
>   ...
>   var personForOffice = {
>     name: 'name',
>     officeName: 'office name'
>   };
>   ```
> * If you get compilation error which says that a property does not exist on a particular type, go through the type definitions of the type and do a type casting if required. For instance:
>   ```
>   var checkMismatch = function(searchQuery) {
>     var isMismatch = true;
>     $('.oppia-search-bar-input').each(function(index) {
>       if ((<string>$(this).val()).trim() === searchQuery) {
>         isMismatch = false;
>       }
>     });
>     return isMismatch;
>   };
>   ```
>   Here `$(this).val()` is type casted to a string by using `<string>$(this).val()`
>   If we do not use a typecast, typescript will give a error `Property 'trim' does not exist on type 'string | number | string[]'` since val can be a string or a number or a string array. So, to use trim we specifically need it as a string.
>   In many cases, you may also need to typecast to `<any>` first and then to the desired type. For example, for converting a number to string, you will need `<string><any>` because neither type sufficiently overlaps with the other.
>   You can add a new custom type definitions if type casting is not possible. In the file `typings/custom-element-defs.d.ts`, we add a new property to `HTMLElement` by adding a custom type defintion. In this type casting cannot be used, since we are adding a new property to the existing type instead of changing it to some other type.
> 
> ### Karma test specific guidelines
> * Use `angular.mock.module` instead of `module` since the typings for angular-mocks does not support the usage of module.
> * Use `angular.mock.inject` instead of `inject` to maintain a consistent behaviour.
> 
> ### When to add custom type defintions to the typings folder?
> * If you find a missing property in a typings package, create an issue [here](https://github.com/DefinitelyTyped/DefinitelyTyped) and a new file for the custom types with the issue link in the top of the file.
> * If you add a package for which type definitions are not found [here](https://github.com/DefinitelyTyped/DefinitelyTyped), add it to `third-party-defs.d.ts`
> * If you add a new property on window which is not present in typings for window, add it to `custom-window-defs.d.ts`
> * If you add a property on scope defined in a link function, add it to `custom-scope-defs.d.ts` and add a comment specifying the filename for which it is added.
> * Make sure that all files have comments which explain why these custom type defintions are required and additional comments to explain each new added property if required. For example, `typings/custom-scope-defs.d.ts` has a top level comment explaining that the type defintions are needed for properties defined on scope in link function and then there are additional comments with properties added specifying which file they belong to. Go through the existing files and try to follow the same pattern when adding a new file.

Hmm, how about adding methods to objects? I see you use `obj['methodName'] = function() {};`, is this the cleanest way to do so?

nithusha21 commented:
Should local_compiled_js/ and tsc_output_log.txt be ignored by git? My git status seems to continue showing these as untracked after I started the server.

ankita240796 committed:
Resolve conflicts

ankita240796 committed:
Disable lint

ankita240796 committed:
Enable lint

ankita240796 commented:
Hi @brianrodri,
> Hmm, how about adding methods to objects? I see you use obj['methodName'] = function() {};, is this the cleanest way to do so?

No, neither any properties nor methods can be added to a defined object in typescript since it basically overrides the type which you defined earlier. For methods too, you will need to declare them in the definition itself either as null or the method if it is known at the time of object declaration.

Hi @nithusha21,
> Should local_compiled_js/ and tsc_output_log.txt be ignored by git? My git status seems to continue showing these as untracked after I started the server.

They should be ignored by git. It is strange that they are being shown by git status as untracked. It does not happen on my system. I will check once again and let you know.

ankita240796 commented:
Hi @nithusha21, I am unable to reproduce the issue as mentioned by you. Can you please check once again if that is happening, Thanks!

ankita240796 committed:
Try fixing dependency issue

ankita240796 committed:
Try fixing dependency issue

ankita240796 committed:
Revert changes

DubeySandeep commented:
@nithusha21, PTAL!

nithusha21 commented:
Resolved this on chat with Ankita, it works as expected. I had run git status on my feature branch after testing this PR, so that's why it behaved like what I mentioned earlier. She has already added it to gitignore.

ankita240796 commented:
Hi @seanlip, the testing is done and changes are being detected on the dev server. We can get this in once the frontend and e2e test issue is resolved.

seanlip commented:
Sounds good! I believe all dev workflow issues are resolved, actually. Try updating your branch with develop?

/cc @oppia/dev-workflow-team for assistance if that doesn't solve the issue. @oppia/dev-workflow-team -- we need to get this in quickly, so any help you can provide here is appreciated. Thanks!

apb7 commented:
Hi @seanlip, I have restarted all the failing tests. This should probably fix them. Thanks!

nithusha21 commented:
Should we merge this PR (before more conflicts happen?) Or do we want to wait on all codeowners?

seanlip commented:
I'm comfortable with merging it. Let me give it a quick once-over and I'll do it. Thanks!

seanlip commented:
OK, it's in! Probably want to give an update to oppia-dev@ :)
